### PR TITLE
HierarchicalOptimization added

### PR DESCRIPTION
**Pull request by LoosC**
added the logLikelihood for the hierarchical optimization and three examples where standard and hierarchical optimization are compared

### CODE REVIEW

File: @HOOptions/HOOptions.m [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+            %   OPTS = HOOOptions(OLDOPTS, PARAM, VAL, ...) creates a copy of OLDOPTS with
+            %   the named parameters altered with the specified value
+            %
+            %   Note to see the parameters, check the
+            %   documentation page for HOOptions
+            %
+            % Parameters:
+            %   varargin:
+            
+            % adapted from SolverOptions
```
Comment (FFroehlich) [inline @HOOptions/HOOptions.m]: see https://github.com/ICB-DCM/PESTO/issues/128

File: @HOOptions/HOOptions.m [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+                    end
+                end
+            end
+            
+            % Add required subclasses
+            obj.HOObservables = HOObservablesOptions();
+            
+        end
+        
+        function new = copy(this)
```
Comment (FFroehlich) [inline @HOOptions/HOOptions.m]: we do not need a copy operator for non-handle classes. Please also delete this where copied it from (this is now the third time I have to complain about this in a PR :().

File: @HOOptions/HOOptions.m [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+            if (strcmp(value, 'normal') || strcmp(value, 'laplace'))
+                this.distribution = value;
+            else
+                error('HOOptions.distribution must be set to either "normal" or "laplace".');
+            end
+        end
+        
+        function set.scale(this, value)
+            if ~isequal(this.n_obs,length(value))
+                error('HOOptions.scale must have the dimension 1xn_obs');
```
Comment (FFroehlich) [inline @HOOptions/HOOptions.m]: ~dimension 1xn_obs~ length n_obs

File: @HOOptions/HOOptions.m [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+                    this.scale(i) = value(i);
+                else
+                    error('HOOptions.scale must have entries "lin", "log", or "log10"');
+                end
+            end
+        end
+        
+        function set.noise(this, value)
+            if ~isequal(this.n_obs,length(value))
+                error('HOOptions.scaling must have the dimension 1xn_obs');
```
Comment (FFroehlich) [inline @HOOptions/HOOptions.m]: ~dimension 1xn_obs~ length n_obs

File: @HOOptions/HOOptions.m (hunk 3 above)
Comment (FFroehlich) [reply @HOOptions/HOOptions.m]: also length(x) == max(size(X)), so the check you are doing here may fail to catch a problem if value is a matrix.

File: examples/HierarchicalOptimization_examples/BachmannJakStat/loadBachmannParameters.m [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            'offset_pJAK2_long'
+            'offset_pJAK2_shp1oe'
+            'offset_pJAK2_socs3oe'
+            'offset_pSTAT5_actd'
+            'offset_pSTAT5_cisoe'
+            'offset_pSTAT5_conc'
+            'offset_pSTAT5_long'
+            'offset_pSTAT5_shp1oe'
+            'offset_pSTAT5_socs3oe'%;
+            %'sd_pSTAT5_socs3oe'
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/BachmannJakStat/loadBachmannParameters.m]: remove?

File: examples/HierarchicalOptimization_examples/BachmannJakStat/loadBachmannParameters.m [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            'offset_pJAK2_fine'
+            'offset_pJAK2_long'
+            'offset_pJAK2_shp1oe'
+            'offset_pJAK2_socs3oe'
+            'offset_pSTAT5_actd'
+            'offset_pSTAT5_cisoe'
+            'offset_pSTAT5_conc'
+            'offset_pSTAT5_long'
+            'offset_pSTAT5_shp1oe'
+            'offset_pSTAT5_socs3oe'
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/BachmannJakStat/loadBachmannParameters.m]: until here, all names match. probably better readable (and less error prone) if you always set these values and append the scaling and sd parameters and if the hierarchical approach is chosen

File: examples/HierarchicalOptimization_examples/BachmannJakStat/logLikelihood_BachmannJakStat.m [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            options.ami.sx0 = D(cond).sinit(xi,D(cond).u);
+        end
+        if cond == 13 || cond == 14
+            sol(cond) = simulate_BachmannJakStat_SHP1oe(D(cond).t,xi(1:27),D(cond).u,[],options.ami);
+        else
+            sol(cond) = simulate_BachmannJakStat(D(cond).t,xi(1:27),D(cond).u,[],options.ami);
+        end
+        
+        if sol(cond).status < 0
+            varargout{1} = nan;
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/BachmannJakStat/logLikelihood_BachmannJakStat.m]: replacing l44-51 by `assert(sol(cond).status>=0)` will do the same thing ;)

File: examples/HierarchicalOptimization_examples/BachmannJakStat/logLikelihood_BachmannJakStat.m [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+            else
+                y_ch = bsxfun(@minus,log10(D(cond).my),log10(sol(cond).y));
+            end
+            switch options.llh.distribution
+                case 'normal'
+                    logL = logL - 0.5*(sum(sum(nansum(bsxfun(@times,~isnan(D(cond).my),...
+                        log(2*pi*sigma2))+bsxfun(@rdivide,bsxfun(@power,y_ch,2),sigma2),1),3),2));
+                    if nargout > 1
+                        if cond == 3 && linRNAflag
+                            iy = [12:17];
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/BachmannJakStat/logLikelihood_BachmannJakStat.m]: please simplify the following lines and add a bit of explanation what is actually being computed

tips: 
```
dlogL = dlogL - 0.5*(permute(sum(sum(nansum(...
...
,1),4),2),[3,2,1]));
```
appears in all cases so carry that out once and only compute the argument. this is
```
bsxfun(@times,permute(bsxfun(@rdivide,(1-bsxfun(@rdivide,...
                                 bsxfun(@power,y_ch,2),sigma2)),sigma2),[1,2,4,3]),dsigma2) -...
                                 bsxfun(@times,1/log(10)*permute(2*bsxfun(@rdivide,y_ch,sigma2),[1,2,4,3]),...
                                 bsxfun(@rdivide,sol(cond).sy,sol(cond).y))...
```
in most cases but sometimes it is `bsxfun(@rdivide,sol(cond).sy(:,iy,:),sol(cond).y(:,iy)))` and sometimes `sol(cond).sy(:,iy,:)`. If you make this more explicit, the code will be better readable and its harder to make mistakes by missing individual expressions in this jungle of expressions.

File: examples/HierarchicalOptimization_examples/BachmannJakStat/logLikelihood_BachmannJakStat.m [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+                                bsxfun(@rdivide,sol(cond).sy,sol(cond).y))...
+                                ,1),4),2),[3,2,1]));
+                        end
+                    end
+                case 'laplace'
+                    logL = logL - sum(sum(nansum(bsxfun(@times,~isnan(D(cond).my),log(2*sigma2))+...
+                        bsxfun(@rdivide,abs(y_ch),sigma2),1),3),2);
+                    
+                    if nargout > 1
+                        if cond == 3 && linRNAflag
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/BachmannJakStat/logLikelihood_BachmannJakStat.m]: same here. bonus points if you can directly encode the differences between normal and laplace. something like `bsxfun(@power,y_ch(:,iy,:),2),sigma2(:,iy,:)))` for normal and `bsxfun(@rdivide,abs(y_ch(:,iy,:)),sigma2(:,iy,:)))` for laplace?

File: examples/HierarchicalOptimization_examples/BachmannJakStat/models/BachmannJakStat_SHP1oe_syms.m [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+ SOCS3nRNA4*SOCS3RNADelay - SOCS3nRNA5*SOCS3RNADelay ;
+ (SOCS3nRNA5*SOCS3RNADelay*nuc)/cyt - SOCS3RNA*SOCS3RNATurn ;
+ SOCS3RNA*SOCS3Eqc*SOCS3Turn - SOCS3*SOCS3Turn + SOCS3oe*SOCS3Turn*SOCS3EqcOE * SOCS3Eqc] ;
+
+model.sym.x0=sym(zeros(numel(model.sym.xdot),1));
+model.sym.x0(1) = init_EpoRJAK2;
+model.sym.x0(7) = (init_SHP1 * (1 + (SHP1oe * SHP1ProOE))); 
+model.sym.x0(9) = init_STAT5;
+
+%% OBSERVABLES
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/BachmannJakStat/models/BachmannJakStat_SHP1oe_syms.m]: remove part that is commented out?

File: examples/HierarchicalOptimization_examples/BachmannJakStat/models/BachmannJakStat_SHP1oe_syms.m [hunk 11]
```diff
@@ -0,0 +1,159 @@
+function model = Bachmann_JAKSTAT_SHP1oe_syms()
+
+syms t
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/BachmannJakStat/models/BachmannJakStat_SHP1oe_syms.m]: we probably can do something like `model = Bachmann_JAKSTAT_syms()` and then simply apply necessary changes right?

File: examples/HierarchicalOptimization_examples/BachmannJakStat/models/BachmannJakStat_syms.m [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+ SOCS3RNA*SOCS3Eqc*SOCS3Turn - SOCS3*SOCS3Turn + SOCS3oe*SOCS3Turn*SOCS3EqcOE * SOCS3Eqc] ;
+
+model.sym.x0=sym(zeros(numel(model.sym.xdot),1));
+model.sym.x0(1) = init_EpoRJAK2;
+model.sym.x0(7) = init_SHP1; 
+model.sym.x0(9) = init_STAT5;
+
+%% OBSERVABLES
+% 
+% model.sym.y = [1 / init_EpoRJAK2 *  2 * (EpoRpJAK2 + p1EpoRpJAK2 + p2EpoRpJAK2 + p12EpoRpJAK2);
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/BachmannJakStat/models/BachmannJakStat_syms.m]: remove

File: examples/HierarchicalOptimization_examples/RafMekErk/models/RafMekErk_hierarchical_syms.m [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+%% OBSERVABLES
+y = [x(2);x(3)];%...
+%      x(3);...
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/RafMekErk/models/RafMekErk_hierarchical_syms.m]: remove

File: examples/HierarchicalOptimization_examples/RafMekErk/models/RafMekErk_syms.m [hunk 14]
```diff
@@ -0,0 +1,78 @@
+function [model] = RafMekErk_syms()
+
+%% CVODES OPTIONS
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/RafMekErk/models/RafMekErk_syms.m]: remove

File: examples/HierarchicalOptimization_examples/RafMekErk/models/RafMekErk_hierarchical_syms.m [hunk 15]
```diff
@@ -0,0 +1,80 @@
+function [model] = RafMekErk_analytical_syms()
+
+%% CVODES OPTIONS
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/RafMekErk/models/RafMekErk_hierarchical_syms.m]: remove

File: examples/HierarchicalOptimization_examples/RafMekErk/models/RafMekErk_hierarchical_syms.m [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+% % absolute tolerance
+% model.atol = 1e-8;
+% % relative tolerance
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/RafMekErk/models/RafMekErk_hierarchical_syms.m]: also here: `model = RafMekErk_syms()`, then apply changes.

File: examples/HierarchicalOptimization_examples/SmallJakStat/logLikelihood_JakStat.m [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+            end
+    end
+catch error_thrown
+    warning(['Evaluation of likelihood failed. ',error_thrown.message]);
+    lLH = nan;
+    gradlLH = nan(length(xi),1);
+    logL = nan;
+    grad = nan(length(xi),1);
+end
+if sol.status < 0
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/SmallJakStat/logLikelihood_JakStat.m]: replace by assert in try catch statement

File: examples/HierarchicalOptimization_examples/SmallJakStat/models/JakStat_hierarchical_syms.m [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+
+y(1) = offset_pSTAT + 1/init_STAT*(pSTAT + 2*pSTAT_pSTAT);
+y(2) = offset_tSTAT + 1/init_STAT*(STAT + pSTAT + 2*(pSTAT_pSTAT));
+y(3) = u(1);
+
+% SYSTEM STRUCT
+
+model.sym.x = x;
+model.sym.u = u;
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/SmallJakStat/models/JakStat_hierarchical_syms.m]: u does not have any effect

File: examples/HierarchicalOptimization_examples/SmallJakStat/models/JakStat_hierarchical_syms.m [hunk 19]
```diff
@@ -0,0 +1,69 @@
+function [model] = JakStat_hierarchical_syms()
+
+% CVODES OPTIONS
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/SmallJakStat/models/JakStat_hierarchical_syms.m]: difference between this and `JakStat_syms()`?

File: examples/HierarchicalOptimization_examples/RafMekErk/logLikelihood_RafMekErk_standard.m [hunk 20]
```diff
@@ -0,0 +1,171 @@
+ function [varargout] = logLikelihood_RafMekErk_standard(xi,D,options)
+% logLikelihood_RafMekErk_standard() computes the log-likelihood function for
+% the RAF/MEK/ERK model in the standard case of optimization.
+%
+% USAGE:
+% * [logL] = logLikelihood__RafMekErk_standard()
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/RafMekErk/logLikelihood_RafMekErk_standard.m]: you can't call this with no arguments: `(...)`

File: logLikelihoodHierarchical.m [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+
+% Initialization for for scaling and noise parameters
+s = zeros(1,n_y,n_r,n_e); % vector including scaling factors
+noise = zeros(1,n_y,n_r,n_e); % vector including noises
+
+if nargout > 1
+    ds = zeros(1,n_y,n_theta,n_r,n_e);
+end
+
+% if ~checkValidityOptions(D,options)
```
Comment (FFroehlich) [inline logLikelihoodHierarchical.m]: remove?

File: logLikelihoodHierarchical.m [hunk 22]
```diff
@@ -0,0 +1,271 @@
...
+        for ir = 1:n_r
+            noise(:,ind_y,ir,ind_e)=tempnoise(:,:,ir);
+        end
+        
+    end
+end
+
+%% save scaling and noise parameters
+if options.save
+    save('analytical_results.mat','s','noise');
```
Comment (FFroehlich) [inline logLikelihoodHierarchical.m]: i thought we have an option in PESTO that specifies the current project folder. We should save the results in that folder,

File: logLikelihoodHierarchical.m [hunk 23]
```diff
@@ -0,0 +1,271 @@
...
+end
+
+%% COMPUTING lLH, gradlLH, HessianlLH
+switch options.distribution
```
Comment (FFroehlich) [inline logLikelihoodHierarchical.m]: as for logLikelihood_BachmannJakStat, there is a large overlap in the code for laplace/gaussian

File: private/optimalNoise.m [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+        ddnoise = zeros(n_theta,1,n_theta,n_r); %second order derivatives of the variances
+    end
+end
+
+try
+    switch options.noise{iy(1)}
+        case 'multiple'
+            mult_fact_ir = zeros(1,1,n_r);
+            noiseir = zeros(1,1,n_r);
+%             if nargout > 1
```
Comment (FFroehlich) [inline private/optimalNoise.m]: remove

File: private/optimalScaling.m [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+
+try
+    switch scaling
+        case 'absolute'
+            s = ones(1,1,n_r);
+            if nargout > 1
+                ds = zeros(1,1,n_theta,n_r);
+            end
+        case 'multiple'
+            switch options.distribution
```
Comment (FFroehlich) [inline private/optimalScaling.m]: For the normal distribution the only difference between `multiple` and `single` i could spot was whether `s = exp(bsxfun(@rdivide,logmy-logy,multfact));` or `s = bsxfun(@times,ones(1,1,n_r),exp((logmy-logy)/multfact));` was called in the end. I believe here is a lot of room to optimize code. This should make it easier to extend the code to other distributions in the future.

File: private/optimalNoise.m [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+                        noiseir = noiseir + sum(nansum(bsxfun(@power,y_sh,2),1),2);
+                    case {'laplace'}
+                        %computing the optimal scale parameter
+                        noiseir = noiseir + sum(nansum(abs(y_sh),1),2);
+                end
+            end
+            
+            noise = bsxfun(@rdivide,noiseir,mult_fact_ir);
+        case 'single'
+            mult_fact_i = zeros(1,1);
```
Comment (FFroehlich) [inline private/optimalNoise.m]: code for `multiple` and `single` are super similar. Should be possible to collapse both cases with a couple of if statements

File: @HOOptions/HOOptions.m [hunk 27]
```diff
@@ -272,8 +272,8 @@
         end
         
         function set.scale(this, value)
-            if ~isequal(this.n_obs,length(value))
-                error('HOOptions.scale must have the dimension 1xn_obs');
+            if (~isequal(this.n_obs,length(value)) || ~isequal(max(size(value)),length(value)))
```
Comment (FFroehlich) [inline @HOOptions/HOOptions.m]: neee, mach doch einfach ~isequal(this.n_obs,numel(value))

File: examples/HierarchicalOptimization_examples/BachmannJakStat/loadBachmannParameters.m [hunk 28]
```diff
@@ -1,187 +1,126 @@
...
+    'sd_SHP1_au'
+    'sd_SOCS3_abs'
+    'sd_SOCS3_au'
+    'sd_STAT5_abs'
+    'sd_STAT5_au'
+    'sd_pSTAT5_rel'
+    };
+
+if strcmp(approach,'hierarchical')
+    parameters_temp = parameters;
```
Comment (FFroehlich) [inline examples/HierarchicalOptimization_examples/BachmannJakStat/loadBachmannParameters.m]: what about `parameters.name = parameters.name(1:58)`?

Comment (LoosC) [reply examples/HierarchicalOptimization_examples/BachmannJakStat/loadBachmannParameters.m]: guter punkt, hatte das erst probiert aber keine ahnung warum das nicht funktioniert hatte, jetzt passts ;)

**Pull request merged**